### PR TITLE
[SourceKit] Cancel in-flight builds on `editor.close`

### PR DESCRIPTION
**Pull request by hamishknight**
When closing a document, cancel any in-flight builds happening for it.

rdar://127126348

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please build toolchain

File: unittests/SourceKit/SwiftLang/CloseTest.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+    "let x: C = 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8\n";
+}
+
+TEST_F(CloseTest, Cancel) {
+  const char *DocName = "test.swift";
+  auto *Contents = getComplexSourceText();
+  const char *Args[] = {"-parse-as-library"};
+
+  for (auto RemoveCache : {true, false, false}) {
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply unittests/SourceKit/SwiftLang/CloseTest.cpp:159]: Heh, I meant to add a comment explaining but never did. It's to test the prior state for both the ASTProducer being cached and not cached

File: tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp [hunk 2]
```diff
@@ -2444,8 +2455,14 @@ void SwiftLangSupport::editorClose(StringRef Name, bool RemoveCache) {
...
 
-  if (Removed && RemoveCache)
-    Removed->removeCachedAST();
+  if (Removed) {
+    // Cancel any in-flight builds for the given AST.
+    Removed->cancelBuildsForCachedAST();
+
+    // Then remove the cached AST if we've been asked to do so.
+    if (RemoveCache)
+      Removed->removeCachedAST();
```
Comment (bnbarham) [inline tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:2464]: Is there any point to not removing from the cache if we never re-use it anyway? Or is there some case where it is re-used?

File: unittests/SourceKit/SwiftLang/CloseTest.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (hamishknight) [reply unittests/SourceKit/SwiftLang/CloseTest.cpp:161]: Added a comment for this

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
After talking with @ahoppen, updated to use an opt-in flag for this behavior

**Comment by hamishknight** [comment]
@swift-ci please build toolchain

**Comment by hamishknight** [comment]
After further discussion, flipped the default back to `true`

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please build toolchain

**Review by bnbarham — APPROVED** [review]

**Pull request merged**